Allowing rootDocument to be overridden

thisispaul opened the pull request:
This allows the root node to be overridden which fixes a problem where the lib could not generate a unique path due to repeated elements.

I have not added any tests as I am not sure how to pass Elements through the options in the tests.

Related: #17

antonmedv commented on .gitignore:
```diff
@@ -1,3 +1,4 @@
 /node_modules/
 /dist/
+/.idea/
```
All system related ignore rules should not be added into project's .gitignore. Use system git config instead.

thisispaul replied:
Removed

antonmedv commented on src/index.ts:
```diff
@@ -16,6 +16,7 @@ enum Limit {
 
 export type Options = {
   root: Element
+  rootDocument: Element | Document
```
I think extra option is not needed as there is already root option.

antonmedv commented on src/index.ts:
```diff
@@ -48,7 +50,7 @@ export default function (input: Element, options?: Partial<Options>) {
   config = {...defaults, ...options}
 
-  rootDocument = findRootDocument(config.root)
+  rootDocument  = config.rootDocument !== undefined ? config.rootDocument as Element : findRootDocument(config.root)
```
Better to setup correct rootDocument inside findRootDocument. See https://github.com/antonmedv/finder/pull/14 for more details on why this was added in the first place.

thisispaul replied:
The only thing I can see of doing here for this is to just always use the config.root as the root document as it defaults to the `document.body` anyway. Is there a reason you are using the `ownerDocument`instead of just the `document.body` or the element passed in?

All the functions you are calling on `rootDocument` are available on Elements as well.

thisispaul replied:
Actually. Just using the config.root as the `rootDocument` might solve this problem and #14 if i understand #14 correctly. I will push the change I think makes sense and ill see if I can set up a test with an iframe too to see if it works.

antonmedv commented on tsconfig.json:
```diff
@@ -11,7 +11,7 @@
     "noImplicitThis": true,
     "noImplicitAny": false,
-    "strictNullChecks": true,
+    "strictNullChecks": false,
```
Better leave strictNullChecks true.

thisispaul commented:
I have changed it now so it is just using the root option.

One thing to mention. I have changed the `findRootDocument` slightly. I have changed it from a ternery operation and have added this check in:
```
  if (rootNode === defaults.root) {
      return rootNode.ownerDocument as Document
  }
```

This is because the tests were failing if I just returned the rootNode when the default was there as the tests are checking for `head` elements which it cannot get if it is only finding selectors from the body. I don't know how useful it is for this library to generate selectors for head but I didn't want to mess with the tests so just did this.
